### PR TITLE
[IR] Add `samesign` flag to icmp instruction

### PR DESCRIPTION
**Pull request by elhewaty**
Inspired by ﻿https://discourse.llvm.org/t/rfc-signedness-independent-icmps/81423

### CODE REVIEW

File: llvm/include/llvm/IR/InstrTypes.h
Comment (tschuett) [inline llvm/include/llvm/IR/InstrTypes.h:1039]: Why is the base class `CmpInst` and not `ÌCmpInst`?

File: llvm/include/llvm/IR/InstrTypes.h
Comment (tschuett) [inline llvm/include/llvm/IR/InstrTypes.h:1054]: isa + cast -> dyn_cast

File: llvm/include/llvm/IR/InstrTypes.h
Comment (nikic) [reply llvm/include/llvm/IR/InstrTypes.h:1039]: I'd prefer to add hasSameSign/setSameSame directly to ICmpInst, instead of having the PossiblySameSignInst wrapper.

Comment (elhewaty) [reply llvm/include/llvm/IR/InstrTypes.h:1039]: > Why is the base class `CmpInst` and not `ÌCmpInst`?

Oops, I missed this!

Comment (elhewaty) [reply llvm/include/llvm/IR/InstrTypes.h:1039]: > I'd prefer to add hasSameSign/setSameSame directly to ICmpInst, instead of having the PossiblySameSignInst wrapper.

@nikic  Is there a specific reason for this?

File: llvm/docs/LangRef.rst
Comment (dtcxzyw) [inline llvm/docs/LangRef.rst:12313]:
```suggestion
same sign then the result is a :ref:`poison value <poisonvalues>`.
```

File: llvm/include/llvm/IR/InstrTypes.h
Comment (nikic) [reply llvm/include/llvm/IR/InstrTypes.h:1039]: ICmpInst is the only instruction that supports the flag, so adding the methods directly there will reduce the amount of casting that has to be done.

Similar to how nuw/nsw are directly set on TruncInst.

**Comment by elhewaty** [comment]
@nikic @dtcxzyw, these are failed tests, any hints why?
```
Failed Tests (16):
  LLVM :: Assembler/auto_upgrade_nvvm_intrinsics.ll
  LLVM :: Assembler/flags.ll
  LLVM :: Bindings/llvm-c/echo.ll
  LLVM :: Bitcode/compatibility-3.6.ll
  LLVM :: Bitcode/compatibility-3.7.ll
  LLVM :: Bitcode/compatibility-3.8.ll
  LLVM :: Bitcode/compatibility-3.9.ll
  LLVM :: Bitcode/compatibility-4.0.ll
  LLVM :: Bitcode/compatibility-5.0.ll
  LLVM :: Bitcode/compatibility-6.0.ll
  LLVM :: Bitcode/compatibility.ll
  LLVM :: Bitcode/flags.ll
  LLVM :: Bitcode/miscInstructions.3.2.ll
  LLVM :: Feature/basictest.ll
  LLVM :: Feature/recursivetype.ll
  LLVM :: Integer/basictest_bt.ll
```

and here is one failure message that I don't understand:
```
FAIL: LLVM :: Assembler/flags.ll (1 of 53995)
******************** TEST 'LLVM :: Assembler/flags.ll' FAILED ********************
Exit Code: 2

Command Output (stderr):
--
RUN: at line 1: myLLVM/build/bin/llvm-as < myLLVM/llvm-project/llvm/test/Assembler/flags.ll | myLLVM/build/bin/llvm-dis | myLLVM/build/bin/FileCheck myLLVM/llvm-project/llvm/test/Assembler/flags.ll
+ myLLVM/build/bin/llvm-as
+ myLLVM/build/bin/llvm-dis
+ myLLVM/build/bin/FileCheck myLLVM/llvm-project/llvm/test/Assembler/flags.ll
myLLVM/build/bin/llvm-dis: error: Invalid record (Producer: 'LLVM20.0.0git' Reader: 'LLVM 20.0.0git')
FileCheck error: '<stdin>' is empty.
FileCheck command line:  myLLVM/build/bin/FileCheck myLLVM/llvm-project/llvm/test/Assembler/flags.ll

--
```

File: llvm/include/llvm/Bitcode/LLVMBitCodes.h
Comment (nikic) [inline llvm/include/llvm/Bitcode/LLVMBitCodes.h:545]:
```suggestion
enum ICmpOptionalFlags { ICMP_SAME_SIGN = 0 };
```

File: llvm/lib/Bitcode/Reader/BitcodeReader.cpp
Comment (nikic) [inline llvm/lib/Bitcode/Reader/BitcodeReader.cpp:5486]: You can't unconditionally read this, you need to check the record length first. See how the IsFP case for FMF is handled above.

This is the cause of the test failures.

Comment (elhewaty) [reply llvm/lib/Bitcode/Reader/BitcodeReader.cpp:5486]: @nikic  I am working on it now. but I don't fully understand why we should record length.
can you please give a brief explanation of what the record does?

Comment (nikic) [reply llvm/lib/Bitcode/Reader/BitcodeReader.cpp:5486]: The bitcode writer only appends the flags record if there are flags, so the bitcode reader needs to check the number of records to determine whether there are any flags.

Comment (elhewaty) [reply llvm/lib/Bitcode/Reader/BitcodeReader.cpp:5486]: So the record size needs to be greater than the number of operands this means that the record has operands and flags?

```
if (Record.size() > OpNum+1 && (Record[OpNum] & (1 << bitc::ICMP_SAME_SIGN)))
          cast<ICmpInst>(I)->setSameSign();
```

Comment (nikic) [reply llvm/lib/Bitcode/Reader/BitcodeReader.cpp:5486]: Basically yes, but the flags will be at Record[OpNum+1] and also you'll trip the preceding `OpNum+1 != Record.size()` check if you do it like this. Instead you should check this at the same place where FMF is set.

File: llvm/include/llvm/IR/Instructions.h
Comment (tschuett) [reply llvm/include/llvm/IR/Instructions.h:1230]: Do we need the `enum` or can this be `constexpr unsigned SameSigm = 1 << 0;` ?

File: llvm/include/llvm/IR/Instructions.h
Comment (tschuett) [reply llvm/include/llvm/IR/Instructions.h:1229]: Do we need the `, which enables more optimizations` or can we be neutral?

File: llvm/docs/LangRef.rst
Comment (dtcxzyw) [inline llvm/docs/LangRef.rst:12404]: Should we note here we only infer `samesign` on unsigned predicates?

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

BTW, I believe we will encounter some miscompilation bugs as we don't drop poison-generating flags for `icmp` properly in some places. Please give me more time to do some testing before leveraging this flag to do more optimizations.

File: llvm/docs/LangRef.rst
Comment (nikic) [reply llvm/docs/LangRef.rst:12404]: I don't think that this belongs in LangRef.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by elhewaty** [comment]
I will remove it, but I think tha LangRef mentions nuw/nsw, nneg, disjoint ...etc

So what do you think?

**Comment by dtcxzyw** [comment]
> I will remove it, but I think tha LangRef mentions nuw/nsw, nneg, disjoint ...etc
> 
> So what do you think?

You don't need do anything. It is just reply to https://github.com/llvm/llvm-project/pull/111419#discussion_r1800270389.

**Comment by nikic** [comment]
Uh yeah, the way GitHub renders that comment is really confusing...

**Comment by elhewaty** [comment]
Please merge, if ok

**Comment by preames** [comment]
It looks like this missed the update to hasPoisonGeneratingFlags().

**Comment by dtcxzyw** [comment]
> BTW, I believe we will encounter some miscompilation bugs as we don't drop poison-generating flags for icmp properly in some places. Please give me more time to do some testing before leveraging this flag to do more optimizations.

I cannot find more miscompilation bugs with mutation-based fuzzers. Feel free to go ahead :)

**Pull request merged**
